tools/DynamoAddinGenerator/Types.cs

```diff
@@ -139,28 +118,49 @@ public DynamoAddinData(IRevitProduct product, IDynamoInstall latestDynamoInstall
             }
             else
             {
-                ClassName = "Dynamo.Applications.DynamoRevitApp";
-                var assemblyName = "DynamoRevitDS.dll";
-                switch (product.VersionString)
-                {
-                    case "Revit2014":
-                        subfolder = "Revit_2014";
-                        ClassName = "Dynamo.Applications.VersionLoader";
-                        assemblyName = "DynamoRevitVersionSelector.dll";
-                        break;
-                    case "Revit2015":
-                        subfolder = "Revit_2015";
-                        break;
-                    case "Revit2016":
-                        subfolder = "Revit_2016";
-                        break;
-                }
+                ClassName = "Dynamo.Applications.VersionLoader";
+                const string assemblyName = "DynamoRevitVersionSelector.dll";
+
                 AssemblyPath = Path.Combine(latestDynamoInstall.Folder, subfolder, assemblyName);
             }
 
             RevitSubfolder = subfolder;
             AddinPath = Path.Combine(product.AddinsFolder, "Dynamo.addin");
         }
+
+        /// <summary>
+        /// Creates DynamoAddinData to generate addin for given Revit product
+        /// based on latest Dynamo installed on the system.
+        /// </summary>
+        /// <param name="revit">Revit Product for which addin to be generated </param>
+        /// <param name="dynamos">Dynamo products installed on the system</param>
+        /// <param name="dynamoUninstallPath">dynamo path being uninstalled, can be 
+        /// null or empty string</param>
+        /// <returns>DynamoAddinData</returns>
+        public static DynamoAddinData Create(IRevitProduct revit, DynamoProducts dynamos, string dynamoUninstallPath)
+        {
+            //Iterate in reverse order to find the first dynamo that is supported for
+            //this revit product
+            var products = dynamos.Products.Reverse();
+            foreach (var p in products)
+            {
+                //If the current product is being uninstalled, don't generate addin data
+                if (string.Equals(p.InstallLocation, dynamoUninstallPath))
```

Comment:
Not quite sure if it is safe to do string comparison. The following link just for your reference:
http://stackoverflow.com/questions/7344978/verifying-path-equality-with-net